src/libsystemd/sd-resolve/sd-resolve.c

```diff
@@ -239,14 +239,13 @@ static int send_addrinfo_reply(
 
         assert(out_fd >= 0);
 
-        resp = (AddrInfoResponse) {
-                .header.type = RESPONSE_ADDRINFO,
-                .header.id = id,
-                .header.length = sizeof(AddrInfoResponse),
-                .ret = ret,
-                ._errno = _errno,
-                ._h_errno = _h_errno,
-        };
+        /* Do not use structured initializer here. MSan or valgrind do not like it. */
```

Comment:
hmm, but i like the structured initializers!

Maybe an option is to add a macro `memzero_checker()` that is called on `resp` before we initialize statically. This would then zero initialize the variable, but only if valgrind/msan are detected, and would be a NOP otherwise.

valgrind/msan show something that is usually indeed a problem after all: that we send something over a socket that has non-initialized padding. In our case that's totally OK though since we just use it to pass data to our own worker tasks, i.e. the receiver and the sender actually are the same code and trust level...

Reply:
The variables are already initialized with zero. But MSan seems to be complaining when the structured value, whose padding seems not initialized, is substituted.

Reply:
Well, if the uninitialized padding is indeed expected and harmless here, there's always `msan_unpoison()` (or the `memzero_checker()`, as proposed by @poettering), so we don't make the code less readable by adding kludges everywhere.